compiler/codegen/library.cpp

```diff
@@ -0,0 +1,196 @@
+  replace += "--fixpath '$(CHPL_HOME)' $CHPL_HOME < cmd.out.tmp";
+
+  std::string res = runCommand(replace);
+  std::string cleanup = "rm cmd.out.tmp";
+  runCommand(cleanup);
+  return res;
+}
+
+static void openLibraryHelperFile(fileinfo* fi,
+                                  const char* name,
+                                  const char* ext = NULL,
+                                  bool useTmpDir = true);
+static void closeLibraryHelperFile(fileinfo* fi, bool beautifyIt = true);
+
+void codegen_library_makefile() {
+  std::string name = "";
+  int libLength = strlen("lib");
+  bool startsWithLib = strncmp(executableFilename, "lib", libLength) == 0;
+  if (startsWithLib) {
+    name += &executableFilename[libLength];
+  } else {
+    // libname = executableFilename when executableFilename does not start with
+    // "lib"
+    name = executableFilename;
+  }
+  fileinfo makefile;
+  // TODO: alter location to use generated library directory
+  openLibraryHelperFile(&makefile, "Makefile", name.c_str(), false);
+
+  // Save the CHPL_HOME location so it can be used in the other makefile
+  // variables instead of letting them be cluttered with its value
+  std::string cmd = "echo $CHPL_HOME";
+  std::string chplHome = runCommand(cmd);
+  fprintf(makefile.fptr, "CHPL_HOME = %s\n", chplHome.c_str());
+
+  // TODO: compileline --includes-and-defines adds -I. to the list
+  // automatically.  If the library is in a different directory from the
+  // Makefile that is using this one, that won't be sufficient to find the
+  // header file.  I will handle this with the output directory change
+  std::string cflags = getCompilelineOption("cflags");
+  cflags.erase(cflags.length() - 1); // remove trailing newline
+  std::string includes = getCompilelineOption("includes-and-defines");
+  fprintf(makefile.fptr, "CHPL_CFLAGS = %s %s\n",
+          cflags.c_str(),
+          includes.c_str());
+
+  std::string libraries = getCompilelineOption("libraries");
+  std::string libname = "-l";
+  if (startsWithLib) {
+    // libname = "-l<name>" when executableFilename = "lib<name>"
+    libname += name;
+  } else {
+    // libname = executableFilename plus the extension when executableFilename
+    // does not start with "lib"
+    libname = name;
+    libname += getExtension();
+  }
+  // TODO: adjust for different location for the library, see earlier TODO
+  fprintf(makefile.fptr, "CHPL_LDFLAGS = -L. %s %s \n",
+          libname.c_str(),
+          libraries.c_str());
+
+  std::string compiler = getCompilelineOption("compiler");
+  fprintf(makefile.fptr, "CHPL_COMPILER = %s\n", compiler.c_str());
+
+  std::string linker = getCompilelineOption("linker");
+  fprintf(makefile.fptr, "CHPL_LINKER = %s\n", linker.c_str());
+
+  std::string linkerShared = getCompilelineOption("linkershared");
+  fprintf(makefile.fptr, "CHPL_LINKERSHARED = %s", linkerShared.c_str());
+
+  closeLibraryHelperFile(&makefile);
+}
+
+const char* getExtension() {
```

Comment:
Can we name this getOutputExtension or getLibraryExtension or something? It's a little to vague a name to be used in other (non-library-codegen) files.